docs/running-calyx/fud2/scripts.md

```diff
@@ -28,6 +28,63 @@ export const verilog_state = state("verilog", ["sv", "v"]);
 
 These two lines define a `calyx` state and a `verilog` state. The `export` prefix means that these variables will be accessible to other scripts that `import "calyx"`.
 
+Now we will define an operation taking the `calyx` state to the `verilog` state. These operations define functions whose arguments are input files and return values are output values. Their bodies consist of commands that will transform those inputs to those outputs.
+
+```rust,ignore
+// define an op called "calyx_to_verilog" taking a "calyx_state" to a "verilog_state".
+defop calyx_to_verilog(c: calyx_state) >> v: verilog_state {
+    // retrieve a variable from the fud2.toml config
+    let calyx_base = config("calyx.base");
+    // retrieve a variable from the config, or a default value
+    let calyx_exe = config_or("calyx.exe", "${calyx_base}/target/debug/calyx");
+    let args = config_or("calyx.args", "");
+
+    // specify a shell command to turn a calyx file "c" into a verilog file "v"
+    shell("${calyx_exe} -l ${calyx_base} -b verilog ${args} ${c} > ${v});
+}
+```
+
+Counterintuitively, `c`, `v`, `calyx_base`, `calyx_exe`, and `args` do not contain the actual variable values. They contain identifiers which are replaced by the values at runtime. This thwarts the possibility of the different config values or different input file names causing different code paths of a given op to run.
```

Comment:
Not totally sure what this means. I'm guessing it means that if you `print(calyx_base)` you don't get the result from the config file. Instead you get some other value which is replaced at runtime. If I'm understanding correctly, I think it's confusing to say this here. It's not really relevant for writing ops, it's a detail of how this is implemented.

Reply:
I think the main thing we need to convey here is that these `defop` bodies generate Ninja code containing Ninja variable references—actual concrete values (like filenames and configuration strings) are only provided at Ninja evaluation time.

So yeah, as @sgpthomas says, if one were to `print` one of these strings, you would not get a concrete filename. I think the way this is tangibly relevant to op-writers is just that they should not think they can write code that looks inside the value of `args`, for example, uses a Rhai `if` to do different things depending on the configuration.

Reply:
^^^
Exactly. At least for me the intuitive thing is to treat these `defop` and `config` variables like strings containing actual values. This probably is true for others as the old rust and Rhai let the user access actual values (an op I saw actually uses this!).

As a side note:
I tried to avoid referencing generated Ninja in the example docs. I feel like mentioning it is a bit confusing because it kind of forces the reader to understand the whole Ninja generation stuff, which isn't complicated, but requires the user have at least a passing familiarity of Ninja. 

I think a way to clarify this is adding the line
> For example, `print(args)` would print a `$args` instead of the value assigned by the config.

Reply:
@sgpthomas, I tried to also clarify the next sentence to make it feel more relevant to the reader. It was a bit weirdly roundabout before.
> An op cannot take different code paths based on config values or different input/output file names.
Do you think the makes the content and purpose of this digression a bit more clear?

If so, I think I'm going to merge this.

Reply:
Yeah, it’s much more clear now.